handshake.md

```diff
@@ -0,0 +1,583 @@
+specification, including input parameters and references to relevant RFCs,
+under *12. DH functions, cipher functions, and hash functions*.
+
+#### 2.3.2 General operation
+The Noise specification explains the general operation of a Noise protocol in
+*5. Processing rules*. This subsection briefly summarizes the process.
+
+Noise provides relatively high-level functions like `WriteMessage()`, which,
+during the Noise Handshake phase of this protocol will produce the proper
+outgoing binary payloads which can be understood by the Noise implementation on
+the other host.
+
+Similarly, `ReadMessage()` handles parsing and incorporating handshake data
+from the other host. The only responsibility of a Cable Handshake
+implementation is to "ferry" these Noise-produced messages between itself and
+the host with whom it is handshaking: what `WriteMessage()` produces is to be
+sent over the network transport to the other host, and what comes in over the
+network transport is fed into `ReadMessage()`.
+
+This repeats until the handshake is complete, after which two `CipherState`
+objects are returned -- one for encryption and one for decryption -- which
+marks the end of the Noise Handshake, and the beginning of Post-Handshake
+Operation, where these keys can then be used for encrypting and decrypting
+Cable Wire Protocol messages to and from the other host.
+
+The section *5. Processing rules* of the Noise specification goes into more
+detail, and may be worth reading through for implementors.
+
+### 2.4 Cabal key
+A collection of peers sharing subsets of a single dataset is called a *cabal*.
+In other messaging software, this might be called a "server" or "instance". A
+cabal contains a collection of channels, of which users can be members, and
+chat messages may be posted.
+
+Each cabal is identified by a 32-byte key, which is ideally generated using the
+best means available for random number generation on the host system. This is
+called the *cabal key*, and functions as the pre-shared key (`psk0` in the
+aforementioned Noise handshake pattern).
+
+Since the cabal key is mixed into Noise handshake state, every handshake with
+another host under this protocol only permits access to a specific cabal. Only
+a pair of hosts using the same pre-shared key in Noise will be able to
+successfully complete a Noise handshake.
+
+The cabal key effectively acts as a "secret passphrase": only hosts who know
+the key can successfully handshake and then exchange Cable Wire Protocol
+messages. If either party doesn't know the key, Noise will indicate handshake
+failure. This key is only ever used locally and is never sent over the network
+transport. Members of a cabal can share the cabal key over various out-of-band
+means (e.g. other chat programs, written on paper, etc.)
+
+The pre-shared key MUST be mixed into the handshake state as per the rules in
+*9. Pre-shared symmetric keys* of the Noise specification.
+
```

Comment:
It's like @RangerMauve says: you'd need to be listening and doing discovery for each cabal you are interested in.

Reply:
Doesn't that simply make the ports you use part of the shared secret? That might allow me to probe the ports of several machines to find out whether they belong to the same cabals or not.

Reply:
I didn't understand this. Could you rephrase? Are you describing an attack vector? What do you mean by port as part of the shared secret?

Reply:
@hackergrrl Sorry, missed your reply there.

Suppose each cabal was identified only by its key, but there was no canonic information on where (i.e., on which port) peers expect connection attempts for that cabal. Suppose further, there is a peer that is interested in five cabals. That peer probably listens for incoming connections on five different ports. Given that there is no "official" port for each cabal, and they want to keep secret which cabals they belong to, they only advertise "you can connect to my cabals on ports 8000 to 8004". Now, if I come along, interested in only a single cabal, and want to connect to this peer, I have to try out all five ports. And if I'm interested in 20 cabals, none of which the other peer also has (which I don't know though), I make 5*20 connection attempts.

The most obvious way to restore efficiency with the current protocol is to have all members in a cabal agree on a port that they all will use for that cabal. But at that point, the port becomes an identifying secret, just like the actual secret key - if I'm not connecting to the correct port, I cannot join, just like if I didn't know the secret key. And it is a secret that a nosy participant can snoop by probing people which ports they expose (for cable, if that is detectable); the nosy peer can then obtain information on who shares cabals with whom.

Does that clarify things?

Reply:
Ah, ok! I can see your understanding now.

Something not mentioned in this specification, but that is planned, is having a discovery mechanism for finding other members of cabals, which will include both an IP **and** port. This will resolve any uncertainty about what port to find which cabal on.